qwerty_ast/src/typecheck.rs

```diff
@@ -0,0 +1,456 @@
+//! Qwerty typechecker implementation: walks the AST and enforces all typing rules.
+
+use crate::ast::*;
+use crate::error::{TypeError, TypeErrorKind};
+use std::collections::HashMap;
+
+//
+// ─── TYPE ENVIRONMENT ───────────────────────────────────────────────────────────
+//
+
+/// Tracks variable bindings (and potentially functions, quantum registers, etc.)
+#[derive(Debug, Clone)]
+pub struct TypeEnv {
+    vars: HashMap<String, Type>,
```

Comment:
You are cooking.

Another thing that could be useful in this struct is a way to count how many times variables are used, specifically linear variables (nonempty qubit registers). But maybe you will find in your implementation that that does not belong here, I don't know